### PR TITLE
Windows Process API

### PR DESCRIPTION
**Pull request by jan-zajic**
Replace **Fork–exec** Unix technique with **CreateProcess** API on MS Win. As Windows processes don't use the *nix fork idiom, there is no analogue in Windows to forking. Therefore fork **without** exec on Win is **not** implemented (it's almost impossible and not so useful on Windows). 

Personally tested and working.

Little bit problematic because of currently win implementation of Crystal::EventLoop, but works, so will simplify further development of windows fork. (raise "No runnables in scheduler" must be removed, see changes, then wait on Process result using Channel works, otherwise it hang). 

At the same time Signal API reduced to minimal Win supported signals.  Also Signal api is not so useful on Win platform.

Part of #5430.
Existing discussion to signals is here: #7339.

There is lot of necessary magic on Win platform when calling CreateProcessW and related apis..

### CODE REVIEW

**Comment by wontruefree** [comment]
this looks like a lot of work.  Thank you for pushing forward Windows support.

**Comment by straight-shoota** [comment]
This is really awesome. Thank you very much!

It's also quite a lot of code and a huge task to review. The overall design looks good to me. But we'll certainly have to talk about the public API in detail.

We should try to split this feature into smaller bites of self-contained changes, to separate commits and preferably even separate PRs. I assume some parts can be extracted and merged on their own. This makes it easier to review and discuss each individual feature without getting lost.

Skimming over it, I could think of a few ideas:

* lib_c re-orderings
* Auxiliaries such as `to_unsafew` (which is not an ideal name).
* Extract POSIX implementation of Process and Signal to a portable API
* Implement Process and Signal for Win32 (maybe this can be even further broken down?)

**Comment by ysbaddaden** [comment]
Thanks for initiating this!

I think the one thing I dont like are the Signal changes. The only reason we need them with Process is to know and report when a spawned process has terminated on POSIX (SIGCHLD), which is achieved differently on Windows. The sigset is only there because of how we receive/handle signals with the EventLoop and fork/exec semantics; it's not needed on Windows.

I would leave Signal alone for the time being, not abstracting it into System (yet) and obviously not requiring it on win32. This would help to focus the pull request on Process itself for Win32. It only needs Signal on POSIX.

Last but not least, having focused commits, instead of one big commit, as proposed by @straight-shoota would be very appreciated.

**Comment by jan-zajic** [comment]
@ysbaddaden, @straight-shoota thank you for first feedback. I made these changes:
* **removed** completely **Signal changes** from this PR and exclude it on Win platform (@ysbaddaden as it's currently required in Process:Status as well, see commit c704eb4)
* **removed** to_unsafew for now, it will be another PR after merge this (it will unify duplicate code that convert Crystal String to windows wide character string in dir, file, process and so on)
* split remaining stuff to multiple commits

File: src/compiler/crystal/compiler.cr [hunk 1]
```diff
@@ -577,7 +577,10 @@ module Crystal
       status = yield command, args
 
       unless status.success?
-        msg = status.normal_exit? ? "code: #{status.exit_code}" : "signal: #{status.exit_signal} (#{status.exit_signal.value})"
+        msg = "code: #{status.exit_code}"
+        {% unless flag?(:win32) %}
+          msg = status.normal_exit? ? "code: #{status.exit_code}" : "signal: #{status.exit_signal} (#{status.exit_signal.value})"
```
Comment (Sija) [inline src/compiler/crystal/compiler.cr]:
```suggestion
          msg = "signal: #{status.exit_signal} (#{status.exit_signal.value})" unless status.normal_exit?
```

File: src/compiler/crystal/tools/playground/server.cr [hunk 2]
```diff
@@ -175,7 +175,10 @@ module Crystal::Playground
       spawn do
         status = process.wait
         @logger.info "Code execution ended (session=#{@session_key}, tag=#{tag}, filename=#{output_filename})."
-        exit_status = status.normal_exit? ? status.exit_code : status.exit_signal.value
+        exit_status = status.exit_code
+        {% unless flag?(:win32) %}
+          exit_status = status.normal_exit? ? status.exit_code : status.exit_signal.value
```
Comment (Sija) [inline src/compiler/crystal/tools/playground/server.cr]:
```suggestion
          exit_status = status.exit_signal.value unless status.normal_exit?
```

File: src/crystal/system/unix/process.cr [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+require "c/unistd"
+
+class Process
+  private def initialize(@pid)
+    @waitpid = wait(pid)
+    @wait_count = 0
+  end
+
+  def wait(pid : LibC::PidT) : Channel(Int32)
+    return Crystal::SignalChildHandler.wait(pid)
```
Comment (Sija) [inline src/crystal/system/unix/process.cr]: Redundant `return`:

```suggestion
    Crystal::SignalChildHandler.wait(pid)
```

File: src/crystal/system/unix/process.cr [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+  # Returns `true` if the process identified by *pid* is valid for
+  # a currently registered process, `false` otherwise. Note that this
+  # returns `true` for a process in the zombie or similar state.
+  def self.exists_system(pid : Int)
+    ret = LibC.kill(pid, 0)
+    if ret == 0
+      true
+    else
+      return false if Errno.value == Errno::ESRCH
+      raise Errno.new("kill")
```
Comment (Sija) [inline src/crystal/system/unix/process.cr]:
```suggestion
      Errno.new("kill")
```

File: src/crystal/system/unix/process.cr [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+  end
+
+  # Returns `true` if the process identified by *pid* is valid for
+  # a currently registered process, `false` otherwise. Note that this
+  # returns `true` for a process in the zombie or similar state.
+  def self.exists_system(pid : Int)
+    ret = LibC.kill(pid, 0)
+    if ret == 0
+      true
+    else
```
Comment (Sija) [inline src/crystal/system/unix/process.cr]: Please use early return here:

```cr
return true if ret == 0
```

File: src/crystal/system/unix/process.cr [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    kill Signal::KILL
+  end
+
+  # See also: `Process.kill`
+  def kill(sig = Signal::TERM)
+    Process.kill sig, @pid
+  end
+
+  # Sends a *signal* to the processes identified by the given *pids*.
+  def self.kill(signal : Signal, *pids : Int)
```
Comment (Sija) [inline src/crystal/system/unix/process.cr]:
```suggestion
  def self.kill(signal : Signal, *pids : Int) : Nil
```

File: src/crystal/system/unix/process.cr [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+    pids.each do |pid|
+      ret = LibC.kill(pid, signal.value)
+      raise Errno.new("kill") if ret < 0
+    end
+    nil
```
Comment (Sija) [inline src/crystal/system/unix/process.cr]:
```suggestion
```

File: src/crystal/system/unix/process.cr [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+  end
+
+  # Creates a process, executes it
+  def create_and_exec(command : String, args : (Array | Tuple)?, env : Env?, clear_env : Bool, fork_input : IO::FileDescriptor, fork_output : IO::FileDescriptor, fork_error : IO::FileDescriptor, chdir : String?, reader_pipe, writer_pipe)
+    if pid = Process.fork_internal(will_exec: true)
+      pid
```
Comment (Sija) [inline src/crystal/system/unix/process.cr]: Please, use early return:

```suggestion
      return pid
```

File: src/crystal/system/unix/process.cr [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+      end
+    end
+  end
+
+  # Runs the given block inside a new process and
+  # returns a `Process` representing the new child process.
+  def self.fork : Process
+    {% raise("Process fork is unsupported with multithread mode") if flag?(:preview_mt) %}
+    if pid = fork_internal(will_exec: false)
+      new pid
```
Comment (Sija) [inline src/crystal/system/unix/process.cr]: ditto

```suggestion
      return new pid
```

File: src/crystal/system/win32/process.cr [hunk 10]
```diff
@@ -0,0 +1,401 @@
...
+    LibC.GetCurrentThread
+  end
+
+  # Returns `true` if the process identified by *pid* is valid for
+  # a currently registered process, `false` otherwise. Note that this
+  # returns `true` for a process in the zombie or similar state.
+  def self.exists_system(pid : Int)
+    ret = LibC.kill(pid, 0)
+    if ret == 0
+      true
```
Comment (Sija) [inline src/crystal/system/win32/process.cr]: ditto

```cr
return true if ret == 0
```

File: src/crystal/system/win32/process.cr [hunk 11]
```diff
@@ -0,0 +1,401 @@
...
+    else
+      return false if Errno.value == Errno::ESRCH
+      raise Errno.new("kill")
+    end
+  end
+
+  def create_and_exec(command : String, args : (Array | Tuple)?, env : Env?, clear_env : Bool, fork_input : IO::FileDescriptor, fork_output : IO::FileDescriptor, fork_error : IO::FileDescriptor, chdir : String?, reader_pipe, writer_pipe)
+    # like in child process
+    begin
```
Comment (Sija) [inline src/crystal/system/win32/process.cr]: redundant `begin .. end`

File: src/crystal/system/win32/process.cr [hunk 12]
```diff
@@ -0,0 +1,401 @@
...
+      nil,                                                                   # module name
+      to_windows_path(command_line),                                         # command line args
+      nil,                                                                   # Process handle not inheritable
+      nil,                                                                   # Thread handle not inheritable
+      LibC::TRUE,                                                            # Set handle inheritance to TRUE
+      LibC::CREATE_UNICODE_ENVIRONMENT | LibC::EXTENDED_STARTUPINFO_PRESENT, # Use UTF-16 ENV Block
+      Process.create_env_block(env, clear_env),
+      chdir ? to_windows_path(chdir) : nil, # Use chdir or parent's starting directory
+      pointerof(startupinfoex),             # Pointer to STARTUPINFOEX structure
+      pointerof(@pi)    )                   # Pointer to PROCESS_INFORMATION structure
```
Comment (Sija) [inline src/crystal/system/win32/process.cr]:
```suggestion
      pointerof(@pi)                        # Pointer to PROCESS_INFORMATION structure
```

File: src/crystal/system/win32/process.cr [hunk 13]
```diff
@@ -0,0 +1,401 @@
...
+  protected def self.on_exited(context : Void*, isTimeOut : Bool)
+    process = Box(::Process).unbox(context)
+    process.on_exited
+  end
+
+  protected def on_exited
+    @child_descriptors.each do |fd|
+      LibC._close(fd.fd) # ignore fail when already closed by child process
+    end
+    if (LibC.GetExitCodeProcess(@pi.hProcess, out exit_code) == LibC::FALSE)
```
Comment (Sija) [inline src/crystal/system/win32/process.cr]:
```suggestion
    if LibC.GetExitCodeProcess(@pi.hProcess, out exit_code) == LibC::FALSE
```

File: src/crystal/system/win32/process.cr [hunk 14]
```diff
@@ -0,0 +1,401 @@
...
+        raise WinError.new("RegisterWaitForSingleObject")
+      end
+      @handle = wait_handle
+      @pi.dwProcessId.to_i32
+    else
+      raise WinError.new("CreateProcessW")
+    end
+  end
+
+  protected def self.on_exited(context : Void*, isTimeOut : Bool)
```
Comment (Sija) [inline src/crystal/system/win32/process.cr]:
```suggestion
  protected def self.on_exited(context : Void*, _did_time_out : Bool)
```

File: src/crystal/system/win32/process.cr [hunk 15]
```diff
@@ -0,0 +1,401 @@
...
+  # An environment block can contain Unicode characters because we includes CREATE_UNICODE_ENVIRONMENT flag in dwCreationFlags
+  # A Unicode environment block is terminated by four zero bytes: two for the last string, two more to terminate the block.
+  protected def self.create_env_block(env : Env?, clear_env : Bool)
+    final_env : Env = {} of String => String
+    if LibC.CreateEnvironmentBlock(out pointer, nil, LibC::FALSE) == LibC::FALSE
+      raise WinError.new("CreateEnvironmentBlock")
+    end
+    env_block = pointer.as(Pointer(UInt16))
+    begin
+      Crystal::System::Env.parse_env_block(env_block) { |key, val|
```
Comment (Sija) [inline src/crystal/system/win32/process.cr]:
```suggestion
      Crystal::System::Env.parse_env_block(env_block) do |key, val|
```

File: src/crystal/system/win32/process.cr [hunk 16]
```diff
@@ -0,0 +1,401 @@
...
+        final_env[key] = val
+      }
```
Comment (Sija) [inline src/crystal/system/win32/process.cr]:
```suggestion
      end
```

File: src/crystal/system/win32/process.cr [hunk 17]
```diff
@@ -0,0 +1,401 @@
...
+    env.try &.each do |key, val|
+      final_env[key] = val
+    end
+    builder = WinEnvBuilder.new
+    final_env.each do |key, val|
+      add_to_env_block(builder, key, val)
+    end
+    # terminate the block
+    builder.write(0_u16)
+    return builder.buffer
```
Comment (Sija) [inline src/crystal/system/win32/process.cr]:
```suggestion
    builder.buffer
```

File: src/crystal/system/win32/process.cr [hunk 18]
```diff
@@ -0,0 +1,401 @@
...
+    data.process_id = process_id
+    LibC.EnumWindows(->Process.enum_windows_callback(LibC::HWND, LibC::LPARAM), Box.box(data).address)
+    data.window_handle
+  end
+
+  protected def self.enum_windows_callback(handle : LibC::HWND, lparam : LibC::LPARAM) : LibC::BOOL
+    data = Box(FindMainWindowParam).unbox(Pointer(Void).new(lparam))
+    process_id = LibC::DWORD.new(0)
+    LibC.GetWindowThreadProcessId(handle, pointerof(process_id))
+    if (data.process_id != process_id || !is_main_window(handle))
```
Comment (Sija) [inline src/crystal/system/win32/process.cr]:
```suggestion
    if data.process_id != process_id || !is_main_window(handle)
```

File: src/crystal/system/win32/process.cr [hunk 19]
```diff
@@ -0,0 +1,401 @@
...
+      return 1
+    else
+      data.window_handle = handle
+      return 0
+    end
+  end
+
+  protected def self.is_main_window(handle : LibC::HWND) : Bool
+    return LibC.GetWindow(handle, LibC::GW_OWNER).address == 0 && LibC.IsWindowVisible(handle) == 1
```
Comment (Sija) [inline src/crystal/system/win32/process.cr]:
```suggestion
    LibC.GetWindow(handle, LibC::GW_OWNER).address == 0 && LibC.IsWindowVisible(handle) == 1
```

File: src/crystal/system/win32/process.cr [hunk 20]
```diff
@@ -0,0 +1,401 @@
...
+  end
+
+  def self.run_ps(powershell_script : String, env : Env = nil, clear_env : Bool = false, chdir : String? = nil) : String
+    Process.run_ps(powershell_script, env: env, clear_env: clear_env, chdir: chdir) do |proc|
+      proc.output.gets_to_end
+    end
+  end
+
+  private def to_windows_path(path : String) : LibC::LPWSTR
+    return path.check_no_null_byte.to_utf16.to_unsafe
```
Comment (Sija) [inline src/crystal/system/win32/process.cr]:
```suggestion
    path.check_no_null_byte.to_utf16.to_unsafe
```

File: src/crystal/system/win32/process.cr [hunk 21]
```diff
@@ -0,0 +1,401 @@
...
+      count = slice.size
+      new_size = @wchar_size + count
+      if new_size > @capacity
+        resize_to_capacity(Math.pw2ceil(new_size))
+      end
+
+      slice.copy_to(@buffer + @wchar_size, count)
+      @wchar_size += count
+
+      nil
```
Comment (Sija) [inline src/crystal/system/win32/process.cr]:
```suggestion
```

File: src/crystal/system/win32/process.cr [hunk 22]
```diff
@@ -0,0 +1,401 @@
...
+    def write(wchar : UInt16)
+      new_size = @wchar_size + 1
+      if new_size > @capacity
+        resize_to_capacity(Math.pw2ceil(new_size))
+      end
+
+      @buffer[@wchar_size] = wchar
+      @wchar_size += 1
+
+      nil
```
Comment (Sija) [inline src/crystal/system/win32/process.cr]:
```suggestion
```

File: src/crystal/system/win32/process.cr [hunk 23]
```diff
@@ -0,0 +1,401 @@
...
+        resize_to_capacity(Math.pw2ceil(new_size))
+      end
+
+      slice.copy_to(@buffer + @wchar_size, count)
+      @wchar_size += count
+
+      nil
+    end
+
+    def write(wchar : UInt16)
```
Comment (Sija) [inline src/crystal/system/win32/process.cr]:
```suggestion
    def write(wchar : UInt16) : Nil
```

File: src/process.cr [hunk 24]
```diff
@@ -1,60 +1,25 @@
...
-
-  # Sends a *signal* to the processes identified by the given *pids*.
-  def self.kill(signal : Signal, *pids : Int)
-    pids.each do |pid|
-      ret = LibC.kill(pid, signal.value)
-      raise Errno.new("kill") if ret < 0
-    end
-    nil
+  def self.pid
+    return self.pid_system
```
Comment (Sija) [inline src/process.cr]:
```suggestion
    pid_system
```

File: src/process.cr [hunk 25]
```diff
@@ -1,60 +1,25 @@
...
   # returns `true` for a process in the zombie or similar state.
   def self.exists?(pid : Int)
-    ret = LibC.kill(pid, 0)
-    if ret == 0
-      true
-    else
-      return false if Errno.value == Errno::ESRCH
-      raise Errno.new("kill")
-    end
+    return self.exists_system(pid)
```
Comment (Sija) [inline src/process.cr]:
```suggestion
    exists_system(pid)
```

File: src/process.cr [hunk 26]
```diff
@@ -186,18 +80,28 @@ class Process
...
     rescue ex
-      process.kill
+      process.terminate_gracefully
       raise ex
     end
   end
 
+  # Terminate process gracefully
+  def terminate_gracefully
+    return self.terminate_system
```
Comment (Sija) [inline src/process.cr]:
```suggestion
    terminate_system
```

File: src/process.cr [hunk 27]
```diff
@@ -186,18 +80,28 @@ class Process
...
+  end
+
+  # Terminate process immediately (kill)
+  def terminate_immediately
+    return self.kill_system
```
Comment (Sija) [inline src/process.cr]:
```suggestion
    kill_system
```

**Comment by asterite** [comment]
@jan-zajic Only to add to what others already said: this is excellent, thank you so much! With efforts like these Windows support is closer and closer. I'll try to review the commits during these days if I can, though I'll admit I'm a bit clueless regarding OS processes, fork, signals, and of course the Windows API.

But at this point I think it's better to get something working and being able to bootstrap the compiler to Windows. At that point development becomes much easier and we can worry more about the small details or performance matters.

Again, thank you! ❤️❤️❤️

**Comment by straight-shoota** [comment]
It would also be nice if you could add some comments explaining the reasoning behind some changes. For example when some code is excluded on `win32` or entirely removed: Why is this necessary, and is it only a temporary solution that needs improvement later, or won't it ever work/is unnecessary, or are there alternatives?

@Sija Please don't suggest circumstantial style improvements while we still haven't discussed the overall API and implementation. This is just unnecessary noise at the moment.
@jan-zajic Maybe next time submit such a PR as a draft.

**Comment by straight-shoota** [comment]
The most important feature is the public API design for `Process`. There need to be some changes to remove POSIXisms and make it portable.

All documentation for public API should be directly in `src/process.cr`. The documentation needs to be universal as does the behaviour. It should mention platform-specific details, though.
We could consider moving the platform-specific implementation entirely into `Crystal::System::Process` and delegate to that from `Process` (there are already examples for this in the file API). Maybe it would be easier to just reopen `Process` and either delegate to the platform-specific methods or utilize `previous_def` for attaching the documentation. I think the latter is what I'd prefer. It avoids duplicate internal methods.

* `Process.fork` should probably be removed entirely. But that's a different story (see #6421).
* Maybe `Process.chroot` as well.
* Return type for process ids should be Int32.
* I'm not sure about the proper API for sending signals/killing a process

Relevant precursor PR: #6744

**Comment by jan-zajic** [comment]
@straight-shoota, @asterite and all other - i was thinking very much about that and here (#8524) is my proposal how I think it's the best to deal with public platform dependent apis. Please let's discuss it there, before we continue our work on this Process related PR ..

**Comment by straight-shoota** [comment]
Sounds good. So do we agree the system implementation should provide the following interface:
```cr
self.exit_system(status : Int32) : NoReturn
self.pid_system : Int64
self.pgid_system : Int64
self.ppid_system : Int64
self.exist_system?(pid : Int64) : Bool
self.signal_system(pid : Int64, signal : Signal) : Nil
self.terminate_system(pid : Int64) : Nil
self.kill_system(pid : Int64) : Nil
system_close : Nil
create_and_exec(*args) : Int64
self.prepare_shell(command : String, args : Enumerable(String)) : {String, Enumerable(String)}
self.wait(pid : Int64)
```
All these methods should probably be private or protected. Those with `_system` suffix should directly relate to public methods without the suffix.

One issue exists for sending signals to or kill/terminate to arbitrary processes (by process id). On POSIX this works fine one process id but the windows implementations need process handle or thread id. When there are only class methods receiving pid, the windows implementation needs to lookup the handle/thread id and can't re-use the one from an instance. Alternatively, we could implement both class methods receiving pid and instance methods which can use available information. Another solution might be to re-purpose `Process` to also function as a handle for external processes and only have instance methods for signals etc. But that would mean most of the regular API wouldn't work on instances, so it's probably not a good idea.

**Comment by jan-zajic** [comment]
I can implement ppid, but please be aware that pid will be reused by other process if parent terminates and we don't open handle to the parent process. It's different from Linux behavior. 

Regarding kill methods - i'm fine with kill/terminate if we rename existing kill method and you ok with breaking change. I try named them different to just to differentiate from that current Unix kill method which means send any signal and not send kill signal in accordance with Unix kill command.

**Comment by jan-zajic** [comment]
Btw @ysbaddaden i forget to say you that i really don't known about some **GetProcessEntry** windows api, it's only name of the function in go where it is implemented..  NtQueryInformationProcess must be involved. Strictly speaking, the parent process (the process which created the child process) is not actually recorded. InheritedFromUniqueProcessId just gives you the process from which attributes were inherited. This is very rarely a problem, but is controllable from the CreateProcess call, and can be something other than parent.

**Comment by straight-shoota** [comment]
We can surely defer the implementation of `ppid` for win32. It doesn't need to be in this PR.

**Comment by jan-zajic** [comment]
* to api proposal - system_close i change to **close_system** and wait to **wait_system** to be in line with others
* **ppid** - Implemented
* OLD **kill** - renamed to signal, but keeped only in unix/process - if i put it in win process it make dependency to **Signal** which is currently excluded as  @ysbaddaden ask at begining. So keep it only in Unix/process and make another PR on signals later, or i can bring back signal refactoring to this PR.. 
* **pgid** i currently removed from Process api and don't want to give it back - it has same semantic as fork etc - Windows don't have any process groups. In Current Win there is similar concept of **job**, but it is only on demand, ie there is no job for Process unless implicitly created and jobs have string names, not any ID's ..

According to that non-portable apis, as it is last remaining design decision necessary to do:

i currently keep **pgid**, **signal**, **fork** and **chroot** as public methods in file unix/process. They are added to reopened class Process.

We can (please decide per-method):
* **keep** it as it is. (i.e. leave it for later decision on remove..?).
* Create these methods in file /process.cr, rename them to _system in impl and implement them as **raise** in Windows version.
* **remove** them completely.  (and _maybe_ implement them in separate shard).

**Comment by straight-shoota** [comment]
It's okay to not implement the methods in question for win32.
But they should be kept available for now. Since they're public API, they should be in `src/process.cr`. I'd suggest to implement them directly there behind a flag to keep the friction low. That's just meant as an interim solution until we decide later whether and how to port them for win32 or remove them.

Regarding `#signal` we should consider to postpone the rename, or rather move it out of this PR. It's not relevant for win32 implementation.

Could you document the internal interface for `Process`'s system implementation to `src/crystal/system/process.cr`? Just copy https://github.com/crystal-lang/crystal/pull/8518#issuecomment-559215010 plus your modifications. This should be similar to other files in that directory (but without a `Crystal::System` namespace type).

**Comment by asterite** [comment]
Splitting this PR into commits is good, but I think it's better to split this into smaller PRs. Reviewing 31 files changed and 1000 lines of code is a task I can't handle right now, maybe it's the same for others.

Also, because it's so big many discussions are all discussed in a single place and it's not easy to follow.

**Comment by jan-zajic** [comment]
Ok let's make it step by step. Closed in favor to #8536 and #8537. After we stabilize API in #8536 i will make PR's for the rest.

**Comment by RX14** [comment]
I know everyone wants to get stuck into the code as soon as possible, but to prevent this in the future, please open an issue before sending a PR.

We have that in the contribution guide, but it's worth repeating: always make an issue about API changes before sending a PR. Whether you've written the code or not.

**Comment by oprypin** [comment]
I don't think splitting into multiple PRs is viable. Discussion on "later" changes could affect too much in how "earlier" changes would've been submitted. And pull requests cannot be stacked (not without write access to the repository, anyway).

**Comment by oprypin** [comment]
The remainder of this PR merged with master is https://github.com/oprypin/crystal/compare/winapi

I am still planning to try to get that merged, but to be fair I don't quite understand it.

FYI @kubo. If you want to take that instead, let me know in advance.

**Pull request closed without merging**